rust/src/daemon.rs

```diff
@@ -246,6 +254,104 @@ pub(crate) fn deployment_layeredmeta_load(
     deployment_layeredmeta_from_commit(deployment.gobj_rewrap(), commit.gobj_rewrap())
 }
 
+#[context("Loading origin status")]
+pub(crate) fn variant_add_remote_status(
+    mut repo: Pin<&mut crate::FFIOstreeRepo>,
+    refspec: &str,
+    base_checksum: &str,
+    mut dict: Pin<&mut crate::FFIGVariantDict>,
+) -> CxxResult<()> {
+    let repo = &repo.gobj_wrap();
+    let dict = &dict.gobj_wrap();
+
+    let (ostree_remote, _) = ostree::parse_refspec(refspec)?;
+    let mut gpg_verify = false;
+    if let Some(ref ostree_remote) = ostree_remote {
+        match repo.remote_get_gpg_verify(ostree_remote.as_str()) {
+            Ok(b) => gpg_verify = b,
+            Err(err) => {
+                if let Some(ioerr) = err.kind::<gio::IOErrorEnum>() {
+                    // If the remote doesn't exist, let's note that so that status can render it
+                    // specially.
+                    if ioerr == gio::IOErrorEnum::NotFound {
+                        let msg = format!("{}", err);
+                        dict.insert("remote-error", &msg);
+                        return Ok(());
+                    }
+                }
+                return Err(err.into());
+            }
+        };
+    }
+
+    dict.insert("gpg-enabled", &gpg_verify);
+    if !gpg_verify {
+        return Ok(()); // note early return; no need to verify signatures!
+    }
+
+    // there must a remote if GPG checking is enabled
+    let ostree_remote = ostree_remote.unwrap();
+
+    match get_cached_signatures_variant(repo, ostree_remote.as_str(), base_checksum) {
+        Ok(sigs) => dict.insert("signatures", &sigs),
+        Err(err) => {
+            // Somehow, we have a deployment which has gpg-verify=true, but we couldn't verify its
+            // signature. Let's not just bomb out here. We need to return this in the variant so
+            // that `status` can show the appropriate msg.
+            tracing::debug!("failed to get cached signature variant: {}", err);
+        }
+    };
+
+    Ok(())
+}
+
+fn get_cached_signatures_variant(
+    repo: &ostree::Repo,
+    remote: &str,
+    checksum: &str,
+) -> Result<glib::Variant> {
+    std::fs::create_dir_all(RPM_OSTREED_COMMIT_VERIFICATION_CACHE)?;
+    let dir = cap_std::fs::Dir::from_std_file(std::fs::File::open(
+        RPM_OSTREED_COMMIT_VERIFICATION_CACHE,
+    )?);
```

Comment:
```
    // Probably move this to the constant
    let cachepath = "rpm-ostree/gpgcheck-cache";
    let run = cap_std::fs::Dir::open_ambient("/run", cap_std::ambient_authority())?;
    run.create_dir_all(cachepath)?;
    let cachedir = run.open_dir(cachepath)?;
```
or so

Reply:
Ahh cool, I had totally missed the `open_ambient_dir` API. Since it's a "constructor", it feels like it should be closer to the top in the docs.